[Team-19][AOS] [ Linus& 히데 Airbnb 두번째주 화요일 PR ] 날짜 선택 로직 , 가격 RangeBar  구현

HyungMinKang opened the pull request:
## 진행사항
- 체크인, 체크아웃 선택 기능
- 체크인, 체크아웃 선택시 날짜 화면에 표시 +   선택시 UI 변경 구현
- 체크인, 체크아웃 입력값에 따른 버튼 활성화/비활성화 구현
- 가격 선택  RangeBar + Graph 를 오픈소스 라이브러리  이용하여 구현

## 결과
![3번째PR](https://user-images.githubusercontent.com/58967292/171125663-4e0c8650-adda-4cbf-9c06-56b4ddef51e5.gif)

## 추후 진행사항
- 인원 선택 페이지를 Jetpack Compose를 이용하여 구현할 예정입니다.
- 로그인  OAuth를  백엔드팀과 함께 구현할 예정입니다.
-  가격 선택 페이지에   RangeBar+ Graph를 표현하는데 사용한 라이브러리 말고 다른 방법이 없는지 학습해 볼 예정입니다. 

## 질문 
-  커스텀 달력을 실무에서  구현할 때는 어떤식으로 구현하는지 궁금합니다. ( 라이브러리를 쓰셔서 하는지 아니면   직접 커스텀하게  View를 구현하시는지 궁금합니다. )
-  가격 선택 페이지에  RangeBar와 함께  막대 그래프를 표현하기 위해  오픈소스 라이브러리 stfalcon_price_rangebar를 사용하였는데요,   막대 그래프를 그려주는 로직이 복잡해서 이게 올바른 방법인지 궁금합니다.   
``` kotlin
  private fun initChart() {
        val priceMap = getPriceTestData()
        val seekBarEntries = ArrayList<BarEntry>()

        // 선 그래프를 막대 그래프로 변경하는 로직
        for (i in PRICE_MIN_VALUE..PRICE_MAX_VALUE) {
            priceMap[i]?.let {
                seekBarEntries.add(BarEntry(i.toFloat(), it.toFloat()))
                seekBarEntries.add(BarEntry(i + SEEKBAR_VALUE_GAP, it.toFloat()))
                seekBarEntries.add(BarEntry(i + SEEKBAR_VALUE_GAP, SEEKBAR_VACANT_VALUE))
                seekBarEntries.add(BarEntry(i + SEEKBAR_VALUE_GAP + SEEKBAR_VACANT_GAP, SEEKBAR_VACANT_VALUE))
            }
        }
        binding.rbPriceRangeWithChart.setEntries(seekBarEntries)
    }
```

cherrylime69 committed:
feat : 체크인, 체크아웃 날짜를 화면에 표시하기 구현

- 콜백 함수를 통해 선택한 날짜를 ViewModel 에서 각각 분기 처리
- flow 를 통해 분기된 체크인, 체크아웃 날짜 데이터를 BindingAdapter 를 통해 화면에 표시

cherrylime69 committed:
feat : 체크인, 체크아웃 날짜 사이에 있는 날짜 배경색 변경 기능 구현 중
- 각 어댑터의 메서드 호출을 통해 ViewModel 에서 생성한 체크인, 체크아웃 LocalDate 객체 전달

cherrylime69 committed:
feat : 체크인, 체크아웃 날짜 선택 기능 구현 완료
- BindingAdapter를 통해 날짜 선택했을 때 배경색 변경 완료

kang-proto committed:
feat: 날짜 페이지 UI 작업 마무리, 입력값에 따른 버튼 활성화/비활성화 로직 구현

- 체크인/체크아웃 날짜 선택시 UI 변경 로직 완성 ( UI 변경을 위해 drawble 파일 정의)
- 선택가능한 날짜중 가장 첫날짜 UI 변경 로직 적용
- 하단 건너뛰기/ 지우기 버튼 로직 구현 (checkIn/checkOut 값 입력여부에 따라 활성화/비활성화)
- 상단 체크버튼 로직 구현 ( 모든 값 입력시 selected=true로 설정)
- 각 버튼 클릭시 페이지 이동여부는 가격범위 입력페이지 구현후 완성시킬 예정

kang-proto committed:
feat: Navigation 이동 수정

- 기존 테스트를 위해 로그인에서 정보 입력화면으로 넘어갔던 부분을 검색화면에서 넘어가도록 수정
- 검색화면에서 검색결과의 각 아이템에 클릭 리스너를 등록, 클릭시 정보입력 페이지로 이동하도록 구현

kang-proto committed:
feat:  달력 프래그먼트에 내비게이션 로직 등록, 가격 페이지 구현을 위해 라이브러리 그래들에 등록

- 달력 화면에 체크버튼클릭시 가격페이지 이동 로직 완성
- 가격 페이지에 PriceRangeSeekBar 뷰추가

cherrylime69 committed:
Merge pull request #37 from hwicode/feature/custom-calendar-select

[Android][히데&라이너스]  달력 페이지 UI,  선택로직 완성,   가격페이지까지 Navigation 이동 로직 완성

HyungMinKang committed:
Merge pull request #38 from hwicode/android

[Android] 달력 페이지 완성

cherrylime69 committed:
feat : 로그인 화면 Webview로 구현 완료
- 로그인 버튼을 클릭하면 로그인 webview 를 보여주는 프래그먼트로 이동
- Oauth 로직에 따라 로그인 기능을 구현할 예정

cherrylime69 committed:
feat : 가격 그래프 화면 일부 구현

cherrylime69 committed:
feat : 가격 그래프 화면 구현 완료
- 라이브러리를 활용하여 차트를 가진 가격 범위 그래프 구현
- 선 그래프를 막대 그래프로 변환하는 로직 구현
- 각 가격 범위에 맞게 TextView의 Text 변경

cherrylime69 committed:
chore : android root 폴더 생성

cherrylime69 committed:
Merge pull request #44 from hwicode/feature/price-range

[Android] 안드로이드 폴더 생성

cherrylime69 commented:
`chore : android root 폴더 생성` `d83912a`
마지막 커밋은 백앤드 팀과 안드로이드 팀과의 코드를 분리하기 위해 root 디렉토리인 android 폴더를 생성하고 프로젝트를 집어 넣은 커밋입니다. 그래서 해당 커밋의 코드는 리뷰하지 않으셔도 됩니다! 😊

sabgilhun commented on android/app/src/main/java/com/example/airbnb/ui/calendar/CalendarFragment.kt:
```diff
@@ -0,0 +1,113 @@
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        navigator = Navigation.findNavController(view)
+        skipBtn = view.rootView.findViewById<Button>(R.id.btn_information_skip)
+        nextBtn = view.rootView.findViewById(R.id.iBtn_information_next)
```
상위 Fragment의 View를 가져오기 사용하신걸로 보이는데요.

프레그먼트간 or 액티비티와 통신 방법은 아래와 같은 방식들이 있습니다. 
https://developer.android.com/guide/fragments/communicate?hl=ko#viewmodel

살짝 봤을땐 지금 NavHost인 화면이 `InformationInputFragment`인걸로 보이는데, 이걸 Activity로 바꾸면 공유 ViewModel을 통해 쉽게 통신할 수 있을것 같습니다.

필수 수정은 아니고, 이런 방식도 있다고 생각해주세요.

HyungMinKang replied:
아래 버튼 이슈까지 고려해  Activity로 수정하였습니다.   버튼 클릭 리스너에 대한 처리를  Activity에서 구현하고, 활성화 비활성화 여부를 판단할 flag를   공유 ViewModel에 정의했습니다.   
Activity 에서 현재 올라와있는 프래그먼트에 대한 정보가 필요해  when 분기문으로 처리했습니다. 
이렇게 처리하다보니  Information Activity의 코드가 분기문이 많아진 느낌이라 괜찮은지 모르겠습니다.   그래서 Fragment와 Actiivity 사이에서  데이터 공유만  viewModel을 통해 하고,  버튼 클릭로직은 이전처럼  Fragment에서 view를 가져와 쓰는 방법이 더 나은지 궁금합니다.

sabgilhun commented on android/app/src/main/java/com/example/airbnb/ui/calendar/CalendarFragment.kt:
```diff
@@ -0,0 +1,113 @@
+            nextBtn.setOnClickListener {
+                navigator.navigate(R.id.action_calendarFragment_to_priceRangeFragment)
+            }
+        }
```
`isSelected == false` 일때 `nextBtn.setOnClickListener(null)` 하는 코드를 넣어야 할것 같아요.

그리고 이게 상위 Fragment의 버튼이라, 다음 화면이 넘어가도 `onClickListener`가 유지됩니다.
그래서 넘어간 다음 다시 `nextBtn` 버튼을 누르면 크래시가 나네요.

넘어가기전에 지워주거나 아예 다른 방법을 찾거나 해야겠네요.

cherrylime69 replied:
네, ViewModel 로 리팩터링하면서 초기화하는 로직을 넣었습니다.

sabgilhun commented on android/app/src/main/java/com/example/airbnb/ui/calendar/CalendarFragment.kt:
```diff
@@ -0,0 +1,113 @@
+
+    private fun addSkipOrEraseButton() {
+        skipBtn.setOnClickListener {
+            if (skipBtn.text ==  getString(R.string.skip_btn_title)) {
+                navigator.navigate(R.id.action_calendarFragment_to_priceRangeFragment)
+            } else {
+                viewModel.eraseSelectedDate()
+                monthAdapter.notifyDataSetChanged()
+            }
+        }
```
위와 여기도 비슷한 이슈가 있겠네요!

cherrylime69 replied:
위와 마찬가지입니다

sabgilhun commented on android/app/src/main/java/com/example/airbnb/ui/calendar/CalendarViewModel.kt:
```diff
@@ -0,0 +1,64 @@
+        this._checkOutStatedFlow.value= null
+    }
+
+    init {
+        makeCalendarData()
+    }
```
Kotlin Convention에 따르면 init block 은 property 선언 다음 입니다!

https://kotlinlang.org/docs/coding-conventions.html#class-layout

cherrylime69 replied:
네, 컨벤션에 맞게 수정하였습니다

sabgilhun commented on android/app/src/main/res/layout/item_calendar_day.xml:
```diff
@@ -3,29 +3,46 @@
     </data>
 
     <androidx.constraintlayout.widget.ConstraintLayout
+        android:id="@+id/cl_calendar_day"
         android:layout_width="match_parent"
-        android:layout_height="wrap_content">
+        android:layout_height="wrap_content"
+        bindDateString="@{today}"
+        checkIn="@{checkIn}"
+        checkOut="@{checkOut}">
```
item_calender_day.xml 구조가 아래와 같이 변경되어 됩니다.

```
// as-is
<ConstraintLayout>
    <TextView/>
</ConstraintLayout>

// to-be
<TextView/>
```

Item의 경우 꼭 layout으로 감싸지 않아도 됩니다~ 오히려 불필요한 계층이 생겨서 안좋습니다.

아시겠지만, BindingAdapter 변경 해주셔야해요.
```
@BindingAdapter("bindDateString", "checkIn", "checkOut")
fun displayCheckOutDate(view: View, currentDateTime: LocalDate?, checkIn: LocalDate?, checkOut: LocalDate?) {
    ...
}
```

HyungMinKang replied:
네 해당 레이아웃 파일 수정후 바인딩 어뎁터도 수정했습니다.

cherrylime69 replied:
네, 수정하였습니다!

sabgilhun commented on android/app/src/main/java/com/example/airbnb/ui/priceRange/PriceRangeFragment.kt:
```diff
@@ -0,0 +1,93 @@
+class PriceRangeFragment : Fragment() {
+
+    lateinit var binding: FragmentPriceRangeBinding
+    private val formatter = DecimalFormat("#,###")
```
formatter를 class property로 빼두고 재사용하는 것 좋네요 👍

cherrylime69 replied:
감사합니다 😁

sabgilhun reviewed (commented):
히데, 라이너스, 안녕하세요~

캘린더/요금 범위 지정 로직들이 쉽지 않아보였는데, 잘 구현하신것 같아요. (따봉) 수고하셨습니다!

## 질문 답변
### 커스텀 달력 구현 방식
보통 앱 전반적인 톤앤매너를 지키기 위해 커스텀하는 것이 일반적일것 같네요. 앱내에서 많이 사용될 수 있는 위젯일수록 장기적으로 편하긴 합니다 ㅎㅎ

### stfalcon price rangebar 라이브러리 사용법
해당 라이브러리 깃헙 페이즈를 봐도 그렇고, 잘 작성하신 것 같아요!?

차트 라이브러리 중에 유명했던게 `MPChart`라고 있긴한데요. 지금 고르신게 빠르게 개발하기엔 딱인것 같네요.
(https://github.com/PhilJay/MPAndroidChart)

sabgilhun commented on android/app/src/main/java/com/example/airbnb/ui/priceRange/PriceRangeFragment.kt:
```diff
@@ -0,0 +1,93 @@
+        val seekBarEntries = ArrayList<BarEntry>()
+
+        // 선 그래프를 막대 그래프로 변경하는 로직
+        for (i in PRICE_MIN_VALUE..PRICE_MAX_VALUE) {
+            priceMap[i]?.let {
+                seekBarEntries.add(BarEntry(i.toFloat(), it.toFloat()))
+                seekBarEntries.add(BarEntry(i + SEEKBAR_VALUE_GAP, it.toFloat()))
+                seekBarEntries.add(BarEntry(i + SEEKBAR_VALUE_GAP, SEEKBAR_VACANT_VALUE))
+                seekBarEntries.add(BarEntry(i + SEEKBAR_VALUE_GAP + SEEKBAR_VACANT_GAP, SEEKBAR_VACANT_VALUE))
+            }
+        }
+        binding.rbPriceRangeWithChart.setEntries(seekBarEntries)
```
const 사용하여 보는 입장에서 더 잘 이해됐던것 같아요. 최고 👍 

const value를 전역적으로 사용하지 않는 다면 아래처럼도 사용할 수 있어요.

```
class PriceRangeFragment {
    ...
    companion object {
        private const val PRICE_MAX_VALUE = 20
    }
}

cherrylime69 replied:
네, 추후에 전역적으로 사용하지 않는다면 companion object 로 바꾸도록 하겠습니다.

kang-proto committed:
fix:  2추차 첫번쨰 PR 리뷰 반영

- InformationInputFragment -> Activity로 변경
- 넘어가기, 지우기,  checked button 로직을 InformationActivity에서 처리
- 각 버튼의 활성화 여부를 flag로 ViewModel에서 저장
- item_day의 레이아웃 수정 + bindingAdapter 수정
- viewModel init블록 위치 수정

cherrylime69 committed:
Merge pull request #55 from hwicode/fix/secondWeekFirstReview

[Android] fix:  2추차 첫번쨰 PR 리뷰 반영
